src/main/java/com/englishtown/bitbucket/hook/MirrorBucketProcessor.java

```diff
@@ -0,0 +1,147 @@
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.time.Duration;
+import java.util.List;
+import java.util.Locale;
+
+import static com.englishtown.bitbucket.hook.MirrorRepositoryHook.PROP_PREFIX;
+
+public class MirrorBucketProcessor implements BucketProcessor<MirrorRequest> {
+
+    static final String PROP_TIMEOUT = PROP_PREFIX + "timeout";
+
+    private static final String DEFAULT_REFSPEC = "+refs/heads/*:refs/heads/*";
+
+    private static final Logger log = LoggerFactory.getLogger(MirrorBucketProcessor.class);
+
+    private final I18nService i18nService;
+    private final PasswordEncryptor passwordEncryptor;
+    private final RepositoryService repositoryService;
+    private final ScmService scmService;
+    private final SecurityService securityService;
+    private final Duration timeout;
+
+    public MirrorBucketProcessor(I18nService i18nService, PasswordEncryptor passwordEncryptor,
+                                 ApplicationPropertiesService propertiesService, RepositoryService repositoryService,
+                                 ScmService scmService, SecurityService securityService) {
+        this.i18nService = i18nService;
+        this.passwordEncryptor = passwordEncryptor;
+        this.repositoryService = repositoryService;
+        this.scmService = scmService;
+        this.securityService = securityService;
+
+        timeout = Duration.ofSeconds(propertiesService.getPluginProperty(PROP_TIMEOUT, 120L));
+    }
+
+    @Override
+    public void process(@Nonnull String key, @Nonnull List<MirrorRequest> requests) {
+        if (requests.isEmpty()) {
+            return;
+        }
+        // Every request is for the same mirror URL, and the same repository ID. In case the
+        // settings (e.g. username/password) have been changed since the first request was
+        // queued, we process the _last_ request in the list. Since mirroring pushes all of
+        // the configured refspecs, any single request should roll up changes from any number
+        // of requests
+        MirrorRequest request = requests.get(requests.size() - 1);
+
+        securityService.withPermission(Permission.REPO_READ, "Mirror changes")
+                .call(() -> {
+                    Repository repository = repositoryService.getById(request.getRepositoryId());
+                    if (repository == null) {
+                        log.debug("{}: Repository has been deleted", request.getRepositoryId());
+                        return null;
+                    }
+                    if (repositoryService.isEmpty(repository)) {
+                        log.debug("{}: The repository is empty", repository);
+                        return null;
+                    }
+                    runMirrorCommand(request.getSettings(), repository);
+
+                    return null;
+                });
+    }
+
+    private void runMirrorCommand(MirrorSettings settings, Repository repository) {
```

Comment:
This no longer needs to see to its own retries. The `BucketedExecutor` will catch/log the exception when it fails and retry up to the number of attempts configured when the executor was created.